### DISCUSSION TITLE
RFC: Example service implementations

### DISCUSSION THREAD

**Discussion by mattpocock** [mattpocock/xstate-catalogue#3] [Ideas]
**TL;DR:** Allow for browsing example implementations along with the machines. For example, the Authentication example might have an example for Auth0, Cognito etc.

## How this might work

In `./lib/machines/implementations/authentication/aws-amplify.ts`

```ts
import { MachineOptions, Sender } from 'xstate';
import {
  AuthenticationMachineContext,
  AuthenticationMachineEvent,
} from '../authentication.machine';
import Auth from '@aws-amplify/Auth';

const options: Partial<
  MachineOptions<AuthenticationMachineContext, AuthenticationMachineEvent>
> = {
  services: {
    checkIfLoggedIn: () => async (send: Sender<AuthenticationMachineEvent>) => {
      try {
        const _user = await Auth.currentAuthenticatedUser();
        send({
          type: 'REPORT_IS_LOGGED_IN',
          userDetails: _user,
        });
      } catch (e) {
        send({
          type: 'REPORT_IS_LOGGED_OUT',
        });
      }
    },
  },
};

export default options;
```

In `./lib/metadata.ts`

```ts
authentication: {
  title: 'Authentication',
  icon: 'LockOpenOutlined',
  version: '0.0.0',
  implementations: [
    {
      id: 'aws-amplify',
      version: '0.1.0',
    },
  ],
},
```

In `./lib/machines/authentication.mdx`:

```mdx
<Implementations.Wrapper>
  <Implementations.Item id="aws-amplify" title="AWS Amplify">

Description, including markdown, here...

  </Implementations.Item>
</Implementations.Wrapper>
```

## Benefits

 - Example implementations would deepen the understanding of the machine.
 - The catalogue becomes even more useful! These examples would be copy/paste-able too.

**Comment by chrisribe** [comment]
Example implementations! They would deepen the understanding of the machines for sure.

One issue right now is there are only the machines and the theory but not the actual implementation that you can mess with.
So on initial review, it seems way more cryptic and does not show off how it could simplify an existing codebase.
